compose/config/serialize.py

```diff
@@ -24,30 +24,34 @@ def serialize_dict_type(dumper, data):
 
 
 def serialize_string(dumper, data):
-    """ Ensure boolean-like strings are quoted in the output and escape $ characters """
+    """ Ensure boolean-like strings are quoted in the output """
     representer = dumper.represent_str if six.PY3 else dumper.represent_unicode
 
     if isinstance(data, six.binary_type):
         data = data.decode('utf-8')
 
-    data = data.replace('$', '$$')
-
     if data.lower() in ('y', 'n', 'yes', 'no', 'on', 'off', 'true', 'false'):
         # Empirically only y/n appears to be an issue, but this might change
         # depending on which PyYaml version is being used. Err on safe side.
         return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='"')
     return representer(data)
 
 
+def serialize_string_escape_dollar(dumper, data):
+    """ Ensure boolean-like strings are quoted in the output and escape $ characters """
+    data = data.replace('$', '$$')
+    return serialize_string(dumper, data)
+
+
 yaml.SafeDumper.add_representer(types.MountSpec, serialize_dict_type)
 yaml.SafeDumper.add_representer(types.VolumeFromSpec, serialize_config_type)
 yaml.SafeDumper.add_representer(types.VolumeSpec, serialize_config_type)
 yaml.SafeDumper.add_representer(types.SecurityOpt, serialize_config_type)
 yaml.SafeDumper.add_representer(types.ServiceSecret, serialize_dict_type)
 yaml.SafeDumper.add_representer(types.ServiceConfig, serialize_dict_type)
 yaml.SafeDumper.add_representer(types.ServicePort, serialize_dict_type)
-yaml.SafeDumper.add_representer(str, serialize_string)
-yaml.SafeDumper.add_representer(six.text_type, serialize_string)
+yaml.SafeDumper.add_representer(str, serialize_string_escape_dollar)
+yaml.SafeDumper.add_representer(six.text_type, serialize_string_escape_dollar)
```

Comment:
Hi @shin- , sorry for being quiet, extermely busy at work at the moment. All your requests are reasonable and when time allows I will accommodate them. The only thing I'm not sure is the representers right here, you're setting them outside of any function call and which one is used is decided only later based on a command line flag (or its lack). Looking into the `yaml` package I don't see any clearly defined API for manipulating the representers other than `add_representer`. The method is:

```python
@classmethod
def add_representer(cls, data_type, representer):
    if not 'yaml_representers' in cls.__dict__:
        cls.yaml_representers = cls.yaml_representers.copy()
    cls.yaml_representers[data_type] = representer
```

So in theory `add_representer` is OK because it overwrites the dict key. Another option would be to work directly with `yaml_representers`, in case we don't want to rely on `add_representer`s internals. I guess these are unlikely to change though.

Would it be possible to actually always create a dumper and use that instead? So we don't have to fiddle with a singleton's state.

Reply:
Sorry for the late response - holiday season and all that :)
I would just remove these two lines here and amend the change in `serialize_config` to something like this instead:
```python
def serialize_config(config, image_digests=None, escape_dollar=True):
  if escape_dollar:
    yaml.SafeDumper.add_representer(str, serialize_string_escape_dollar)
    yaml.SafeDumper.add_representer(six.text_type, serialize_string_escape_dollar)
  else:
    yaml.SafeDumper.add_representer(str, serialize_string)
    yaml.SafeDumper.add_representer(six.text_type, serialize_string)
```

That way, it's immediately clear what happens in both cases, and if the implementation of `add_representer` changes in a later version of the library (or if we end up switching to the ruamel.yaml package) sometime in the future we don't run into weird side-effects that'll be hard to trace back.

Hope that makes sense - thanks for your help!

Reply:
Makes sense, I just wasn't sure if there's a code path that uses the dumper but doesn't call `serialize_config` where this setup would leave that code path without any representers. I'm on holiday for 2 more weeks and am busy otherwise but hope to kick this off till the end of the month.